pkgs/build-support/zig/build-zig-package/default.nix

```diff
@@ -0,0 +1,44 @@
+{ lib, stdenv, zig }:
+argsFun:
+
+let
+  wrapDerivation = f:
+    stdenv.mkDerivation (finalAttrs:
+      f (lib.toFunction argsFun finalAttrs)
+    );
+in
+wrapDerivation (
+  { strictDeps ? true
+  , nativeBuildInputs ? [ ]
+  , meta ? { }
+  , ...
+  }@attrs:
+
+  attrs // {
+    inherit strictDeps;
+    nativeBuildInputs = [ zig ] ++ nativeBuildInputs;
+
+    buildPhase = attrs.buildPhase or ''
+      runHook preBuild
+      export ZIG_GLOBAL_CACHE_DIR=$(mktemp -d)
+      zig build -Drelease-safe -Dcpu=baseline $zigBuildFlags
+      runHook postBuild
+    '';
+
+    checkPhase = attrs.checkPhase or ''
+      runHook preCheck
+      zig build test
+      runHook postCheck
+    '';
+
+    installPhase = attrs.installPhase or ''
+      runHook preInstall
+      zig build -Drelease-safe -Dcpu=baseline $zigBuildFlags --prefix $out install
+      runHook postInstall
+    '';
+
+    meta = {
+      inherit (zig.meta) platforms;
+    } // meta;
+  }
```

Comment:
This function has the shape of an overlay, if you bring `finalAttrs` into scope.
 - this will be helpful if you have to need the overlay to be aware of the final attrs (at some point)
 - you can document it as such

Specifically you can then say:

`buildZigPackage` invokes `mkDerivation` with your package definition and then overlays it, to add its defaults.
Hence the layers are applied in the following order:
1. your argument to `buildZigPackage`
2. zig defaults
3. anything changed afterwards with `overrideAttrs`
4. `mkDerivation` logic, such as the addition of `finalAttrs.finalPackage`